src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -0,0 +1,222 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.util.LRUCache;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
+import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;
+import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
+import software.amazon.awssdk.services.secretsmanager.model.SecretsManagerException;
+import software.amazon.awssdk.utils.Pair;
+
+import java.sql.SQLException;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+public class AWSSecretsManagerPlugin implements IConnectionPlugin {
+  static final String SECRET_ID_PROPERTY = "secretsManagerSecretId";
+  static final String REGION_PROPERTY = "secretsManagerRegion";
+  private static final String ERROR_MISSING_DEPENDENCY =
+      "[AWSSecretsManagerPlugin] Required dependency 'AWS Java SDK for AWS Secrets Manager' is not on the classpath";
+  static final String ERROR_GET_SECRETS_FAILED =
+      "[AWSSecretsManagerPlugin] Was not able to either fetch or read the database credentials from AWS Secrets Manager. Ensure the correct secretId and region properties have been provided";
+  static final String ERROR_INITIAL_CONN_FAILED =
+      "[AWSSecretsManagerPlugin] Open initial connection failed with given credentials. Retrieving secrets was not attempted";
+  static final String SQLSTATE_ACCESS_ERROR = "28000";
+  static final LRUCache<Pair<String, Region>, Secret> SECRET_CACHE = new LRUCache<>(100);
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+  private final String secretId;
+  private final Region region;
+  private final SecretsManagerClient secretsManagerClient;
+  private final GetSecretValueRequest getSecretValueRequest;
+
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  static class Secret {
+    @JsonProperty("username")
+    private String username;
+    @JsonProperty("password")
+    private String password;
+
+    Secret() {
+    }
+
+    Secret(String username, String password) {
+      this.username = username;
+      this.password = password;
+    }
+
+    String getUsername() {
+      return this.username;
+    }
+
+    String getPassword() {
+      return this.password;
+    }
+  }
+
+  public AWSSecretsManagerPlugin(
```

Comment:
I see two similar constructors. Can we optimize them in order to avoid code repetition?

Reply:
I believe the second constructor is for testing purposes and should be package private.